### PR TITLE
fix(metrics): Fix conversion of distribution to aggregated histogram

### PR DESCRIPTION
**Pull request by jszwedko**
In https://github.com/vectordotdev/vector/pull/10741 we introduced logic
to normalize distributions into aggregated histograms to avoid holding
onto the samples forever. However, the conversion created "cummulative
buckets" which were then accumulated again in the `prometheus_exporter`
sink.

The rest of the code seems to expect that aggregated histograms, in
Vector, have discrete, rather than cummulative, buckets (e.g.
https://github.com/vectordotdev/vector/blob/0ba0d97f5dd116e37b369f3adc755d53ab74d936/lib/vector-core/src/metrics/ddsketch.rs#L596-L601),
so this fixes the conversion to create discrete buckets.

Also removes handling of distributions in the collector since they've
already been normalized.

Closes: #11180

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

### CODE REVIEW

**Comment by binarylogic** [comment]
We need tests for this, right? It seems like one of those tricky behavior gotchas that could benefit from at least one test.

File: lib/vector-core/src/event/metric.rs [hunk 1]
```diff
@@ -1237,18 +1237,19 @@ fn write_word(fmt: &mut Formatter<'_>, word: &str) -> Result<(), fmt::Error> {
...
+fn samples_to_buckets(samples: &[Sample], buckets: &[f64]) -> (Vec<Bucket>, u32, f64) {
     let mut counts = vec![0; buckets.len()];
     let mut sum = 0.0;
     let mut count = 0;
     for sample in samples {
-        buckets
+        if let Some((i, _)) = buckets
             .iter()
             .enumerate()
             .skip_while(|&(_, b)| *b < sample.value)
```
Comment (tobz) [inline lib/vector-core/src/event/metric.rs:1248]: As an aside, and not something we need to address here: I think what end up being more generically compatible, in terms of ecosystem/interoperation, would be having buckets be "left inclusive/right exclusive".  For example, if we had two buckets -- upper bounds of 1.0 and 2.0, respectively -- we would describe the second bucket as encompassing all values from (and including) 1.0 up to (but not including) 2.0.

This seems to be the norm in a few numercial/statistical libraries and products/services I sampled:  NumPy, `ndarray_stats`, MATLAB, Circonus, and others.

The trouble with this is that Prometheus, in particular, expects things to be left exclusive/right inclusive... and we can't transform between the two schemes without the original samples.  Something for another day, though.

File: src/sinks/prometheus/collector.rs [hunk 2]
```diff
@@ -54,65 +54,8 @@ pub(super) trait MetricCollector {
...
-                        );
-                        self.emit_value(timestamp, name, "_min", statistic.min, tags, None);
-                        self.emit_value(timestamp, name, "_max", statistic.max, tags, None);
-                        self.emit_value(timestamp, name, "_avg", statistic.avg, tags, None);
-                    } else {
-                        self.emit_value(timestamp, name, "_sum", 0.0, tags, None);
-                        self.emit_value(timestamp, name, "_count", 0.0, tags, None);
-                    }
+                MetricValue::Distribution { .. } => {
+                    unreachable!("distributions have been normalized to histograms or sketches")
```
Comment (tobz) [inline src/sinks/prometheus/collector.rs:58]: Hrmph, I'd have to look at the code but I'm not entirely sure we normalize metrics in the same way for both the remote write _and_ exporter sinks.

Overall, I do like that this involves us getting rid of duplicative code, though.

**Comment by jszwedko** [comment]
@tobz I replaced the distribution handling in [c80e8c5](https://github.com/vectordotdev/vector/pull/11231/commits/c80e8c552a6796c6fc38bdd895652fde15717615) including handling accumulating the buckets correctly.

**Review by tobz — APPROVED** [review]
Seems right to me so long as tests pass.

**Pull request merged**